### PR TITLE
Avoid skipping unused shards twice (for query processing stage and read itself)

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Add `ClusterPtr` to the `SelectQueryInfo` and pass `SelectQueryInfo` to the `getQueryProcessingStage()`

### CODE REVIEW

**Comment by azat** [comment]
>Performance — Errors while building the report.

```
Code: 190, e.displayText() = DB::Exception: Arrays passed to arrayMap must have equal size (version 20.10.1.4626)
```

@akuzm have you seen this? (there is the same error in #14221)

**Comment by azat** [comment]
Not significant, but maybe someone can take a look?

**Comment by azat** [comment]
@alexey-milovidov can you take a look? (Maybe there is some public team, with alias, i.e. `@maintainers` that is preferable over personal ping?)

**Comment by alexey-milovidov** [comment]
I will take a look.

PS. All maintainers are supposedly subscribed to events on GitHub.

**Comment by alexey-milovidov** [comment]
> Yandex third-party checks (only for Yandex employees) 

Need manual editing of third-party code, I will do...

File: src/Storages/StorageDistributed.cpp
Comment (alexey-milovidov) [inline src/Storages/StorageDistributed.cpp:440]: Why we cannot drop `const` in SelectQueryInfo to avoid confusion?

Comment (azat) [reply src/Storages/StorageDistributed.cpp]: Did it my analogy with `input_order_info` which had mutable qualifier and for simplicity (to avoid possible conflicts and to make review easier), but agree this should be cleaned up

FWIW, done in a separate commit

**Comment by azat** [comment]
>Integration tests (asan) — fail: 1, passed: 839, error: 0
>Integration tests (release) — fail: 2, passed: 838, error: 0

Does not looks related

**Comment by azat** [comment]
>Need manual editing of third-party code, I will do...

@alexey-milovidov any news? (I'm curious because this patch set can cause conflicts or some new storage will be added and it will break compilation)

**Comment by azat** [comment]
>Conflicting files
>src/Storages/IStorage.h

rebased against upstream/master

**Comment by alexey-milovidov** [comment]
> @alexey-milovidov any news?

No news so far.

**Comment by azat** [comment]
>No news so far.

Then how about split out 56187e694fa47eac710a1f616c782e5a053ced0c ? This will reduce number of conflicts later

**Comment by azat** [comment]
>Then how about split out 56187e6 ? This will reduce number of conflicts later

https://github.com/ClickHouse/ClickHouse/pull/15548

**Comment by azat** [comment]
>No news so far.

@alexey-milovidov so waht is the fate of this PR?
Due to it changes the interface and some internal code uses them and breaks it cannot be applied?

**Comment by azat** [comment]
@alexey-milovidov looks like `Yandex third-party checks` is broken in master, maybe it's time to break it additionally with this changes? :)

**Review by alexey-milovidov — APPROVED** [review]
LGTM

**Comment by alexey-milovidov** [comment]
Editing third-party code should be easy.

> looks like Yandex third-party checks is broken in master

We sent a patch for them in another PR but they did not merge it yet. We should not wait for them.

**Comment by alexey-milovidov** [comment]
I will merge with red "Yandex third-party checks (only for Yandex employees)"

Info if we need to re-apply commits:
```
commit 381a5b5f55d3559576ad30c522c3e3b7b295c24f (HEAD -> groups/clickhouse_sync/ClickHouse_ClickHouse_pull_14731_5, arcadia/groups/clickhouse_sync/ClickHouse_ClickHouse_pull_14731_5)
author: milovidov
date: 2020-11-08T03:37:45+03:00

    Fix "Arcadia", part 2

commit 3a7aca79479c6e95b12a5a9a1502757384ce7792
author: milovidov
date: 2020-11-08T02:38:17+03:00

    Fix "Arcadia"
```

**Pull request merged**
